diag_manager/fms_diag_reduction_methods.F90

```diff
@@ -0,0 +1,102 @@
+  !! If both the beggining and ending indices are present, then field is assumed to have halos.
+  !! If only beggining indices are present, then field is assumed to be a window.
+  !> @par
+  !! There are a number of ways a user could mess up this logic, depending on the combination
+  !! of presence/absence of is,ie,js,je. The checks below should catch improper combinations.
+  pure function check_indices_order(is_in, ie_in, js_in, je_in) &
+  result(error_msg)
+    integer, intent(in), optional :: is_in, ie_in, js_in, je_in !< Indices passed to fms_diag_accept_data()
+    character(len=128) :: error_msg !< An error message used only for testing purpose!!!
+
+    error_msg = ""
+    IF ( PRESENT(ie_in) ) THEN
+      IF ( .NOT.PRESENT(is_in) ) THEN
+        error_msg = 'ie_in present without is_in'
+        return
+      END IF
+      IF ( PRESENT(js_in) .AND. .NOT.PRESENT(je_in) ) THEN
+        error_msg = 'is_in and ie_in present, but js_in present without je_in'
+        return
+      END IF
+    END IF
+
+    IF ( PRESENT(je_in) ) THEN
+      IF ( .NOT.PRESENT(js_in) ) THEN
+        error_msg = 'je_in present without js_in'
+        return
+      END IF
+      IF ( PRESENT(is_in) .AND. .NOT.PRESENT(ie_in) ) THEN
+        error_msg = 'js_in and je_in present, but is_in present without ie_in'
+        return
+      END IF
+    END IF
+  end function check_indices_order
+
+  !> @brief Sets the logical mask based on mask or rmask
+  !> @return logical mask
+  function init_mask(rmask, mask, field) &
+  result(oor_mask)
+    LOGICAL,  DIMENSION(:,:,:,:), pointer, INTENT(in) :: mask  !< The location of the mask
+    CLASS(*), DIMENSION(:,:,:,:), pointer, INTENT(in) :: rmask !< The masking values
+    CLASS(*), DIMENSION(:,:,:,:),          intent(in) :: field !< Field_data
+
+    logical, allocatable, dimension(:,:,:,:) :: oor_mask !< mask
+
+    ALLOCATE(oor_mask(SIZE(field, 1), SIZE(field, 2), SIZE(field, 3), SIZE(field, 4)))
+    oor_mask = .true.
+
+    if (associated(mask)) then
+      oor_mask = mask
+    elseif (associated(rmask)) then
+      select type (rmask)
+      type is (real(kind=r8_kind))
+        WHERE (rmask < 0.5_r8_kind) oor_mask = .FALSE.
+      type is (real(kind=r4_kind))
+        WHERE (rmask < 0.5_r4_kind) oor_mask = .FALSE.
+      end select
+    endif
+
+  end function init_mask
+
+  !> @brief Sets the weight based on the weight passed into send_data (1.0_r8_kind if the weight is not passed in)
+  !> @return weight to used when averaging
+  pure function set_weight(weight) &
+  result(out_weight)
+    CLASS(*), INTENT(in), OPTIONAL :: weight !< The weight used for averaging
+
+    real(kind=r8_kind) :: out_weight
+
+    out_weight = 1.0_r8_kind
+    if (present(weight)) then
+      select type(weight)
+      type is (real(kind=r8_kind))
+        out_weight = real(weight, kind = r8_kind)
+      type is (real(kind=r4_kind))
+        out_Weight = real(weight, kind = r4_kind)
```

Comment:
You declared `out_Weight` as r8, but here you are making it r4?  Maybe you need `out_Weight` to be class(*) and allocate it to whatever `weight` is?  Then `out_weight = 1.0_r8_kind` would be in the `class default` branch of the `select type`.  Otherwise you should have `kind=r8_kind` on this line I think.

Reply:
`out_weight` is going to be saved as a `r8_kind` and converted to `r4_kind` if needed when doing the mad.

That should be `kind=r8_kind` no `kind=r4_kind`